### ISSUE TITLE
LLK for float32 tilize is truncating small values to 0 that fit within TF32 precision

### ISSUE THREAD

**Issue by yugi957** [tenstorrent/tt-metal#30400]
Feel free to correct the issue if incorrect, but fp32 tilize (which converts to tf32) should be precise enough to handle the below test of .00001, but instead truncates the value to 0 (you can try with larger shapes than scalar as well). Are there intermediate steps in the LLK that would cause this? See below that bf16 passes, where float32 does not.

```
import pytest

import torch

import ttnn

from tests.ttnn.utils_for_testing import assert_with_pcc, check_with_pcc_without_tensor_printout
from models.common.utility_functions import is_grayskull, is_blackhole, torch_random

ttnn_to_torch_dtype = {
    ttnn.float32: torch.float32,
    ttnn.bfloat16: torch.bfloat16,
}


@pytest.mark.parametrize("width", [32, 64])
@pytest.mark.parametrize("val", [0.00001])
@pytest.mark.parametrize(
    "from_to", [(ttnn.ROW_MAJOR_LAYOUT, ttnn.TILE_LAYOUT), (ttnn.TILE_LAYOUT, ttnn.ROW_MAJOR_LAYOUT)]
)
@pytest.mark.parametrize("dtype", [ttnn.float32, ttnn.bfloat16])
def test_to_layout(device, width, from_to, dtype, val):
    from_layout, to_layout = from_to
    torch_input = torch.tensor(val, dtype=ttnn_to_torch_dtype[dtype])

    input_tensor = ttnn.from_torch(
        torch_input,
        dtype=dtype,
        device=device,
        layout=from_layout,
        memory_config=ttnn.DRAM_MEMORY_CONFIG,
    )

    assert input_tensor.layout == from_layout

    output_tensor = ttnn.to_layout(input_tensor, to_layout)
    assert output_tensor.layout == to_layout

    output_tensor = ttnn.to_torch(output_tensor)

    assert_with_pcc(torch_input, output_tensor)
    assert torch.allclose(torch_input, output_tensor)

```

**Comment by ntarafdar** [comment]
@smehtaTT this is a bug for the LLK team not @yugi957 can we re-assign, or do they need more information from us?

**Comment by fahmedchanTT** [comment]
@vmilicevicTT For our review in next sync

**Comment by vmilicevicTT** [comment]
@nvelickovicTT , let's review this during our next sync.  @fahmedchanTT , please make sure this is prioritized on Metal Global Dashboard.  Thanks!

**Comment by ntarafdar** [comment]
what's the status here? @vmilicevicTT , this is blocking a few issues.

**Comment by vmilicevicTT** [comment]
@cmaryanTT , @fahmedchanTT , can you please prioritize this in Metal board along with other P1s that we have?  @ntarafdar , we are not working on this yet due to other priorities, therefore asking @cmaryanTT 's help to prioritize.

**Comment by nvelickovicTT** [comment]
@pmilenkovicTT will take a quick look into this, to see if it can be quickly explained. In general, this code path is fairly well tested, so I would be surprised if it is a bug, rather than misunderstanding of what to expect from HW.
Still, please prioritize on your board accordingly as well @cmaryanTT

**Comment by pmilenkovicTT** [comment]
I took a quick look at this one. 
The behavior seems expected to me from a higher level:
In the unpack stage, we are performing conversion:
```
constexpr std::int32_t unpack_src_format[32] = {
    0 /* Float 32 */ ,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,0,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,
};
constexpr std::int32_t unpack_dst_format[32] = {
    1 /* Float16 */ ,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,1,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,
};
```
I'd expect to have a precision loss here. 

If the full precision of float32 is necessary, we can put that on the map for the LLK team and prioritize. We have a code path for 32 bit datums, as we needed to add support for int32 data type.  

Please prioritize and let us know.

**Comment by nvelickovicTT** [comment]
@yugi957 I see you merged https://github.com/tenstorrent/tt-metal/pull/31843. This is in summary what @pmilenkovicTT pointed out above. That if Dest is in 16bit mode, FP32 format will get unpacked as FP16, while with Dest set to 32bit mode, it should unpack it to TF32.
Did this solve the issue you were seeing?

**Comment by nvelickovicTT** [comment]
Also, there is a rounding effect in SRC regs that rounds the floating point number to a 0, if exponent bits are all 0s. 0.00001 value has all exp bits of 0 in FP16 representation.

The follow up question I don't get here is why did data format inference, infere fp16 as the format, and not bfloat16.

**Comment by cmaryanTT** [comment]
@itarabanTT - there is something wrong at the LLK level here, not using tf32 correctly; but more generally we need a lossless fp32 tilize/untilize path that keeps full fp32 precision.

**Comment by pmilenkovicTT** [comment]
For untilize that's already achievable. For tilize, we can add it to the board and prioritize adding the lossless path.

**Comment by cmaryanTT** [comment]
Existing 32 bit tilize LLK should already be capable of dealing with fp32 tilize, but supposedly there's a Metal check that only uses it for int32. This should be trivial to switch for fp32 and test out.

**Comment by nvelickovicTT** [comment]
@cmaryanTT I mixed this with the transpose implementation. So, here for tilize there is still a small change on LLK side that needs to be done.
And @pmilenkovicTT can do that when he comes back next week, since noting seems to be blocked by this.

**Comment by pmilenkovicTT** [comment]
Hey folks - I am back from the vacation and catching up. 
I ran the test from description on both WH and BH and it's passing on the latest main - I assume it's fixed by [#33904](https://github.com/tenstorrent/tt-metal/pull/33904) - do we know the status of this issue?

**Comment by nvelickovicTT** [comment]
If tf32 underlying issue has been resolved in this case, @cmaryanTT I propose to close this issue and open a separate one for the true fp32 precision LLK implementation.

**Comment by cmaryanTT** [comment]
Tracking the main issue in https://github.com/tenstorrent/tt-metal/issues/35303 (OP) and https://github.com/tenstorrent/tt-metal/issues/35302 (LLK)
